Feature: invitations

machisuji opened the pull request:
WP [#18567](https://community.openproject.org/work_packages/18567)

This PR does a bit more than specified in the work package, but the changes are related.
Will amend the work package with a detailed list of use cases affected or added through this PR.

Main feature: invite users directly via just their email from within a project's members form.

In the course of this the default user creation has been simplified. The form only takes the absolute minimum of necessary information (email, first & last name) and uses the new UserInvitation module for creating the user. An invited user will get an activation mail directly and only then they can choose themselves how to authenticate, i.e. the used password or OmniAuth provider.

machisuji committed:
basic user invitation prototype

machisuji committed:
extract members tab; invite members via email

machisuji committed:
switch to user invitation by default

this changes quite many flows. With this user self registration
is a separate flow to what happens when an admin creates a user.
Users created by admins are 'invited' instead of 'registered'.

machisuji committed:
user invitation polish

machisuji committed:
more polish

machisuji committed:
fixed members index

machisuji committed:
scrap new member modal

machisuji committed:
invited uses should be usable too

We may need to think of a name that includes active, registered
as well as invited users. For now we can think of invited users
as registered ones. Just a bit differently.

machisuji committed:
don't invite existing users; some polish

machisuji committed:
auto focus user select field when adding members

machisuji committed:
more helpful flash messages for project members view

machisuji committed:
you can always add members via email

machisuji committed:
icon for invited users in members view

machisuji committed:
User ⊆ Principal ∧ Principal ⊄ User

I.e. calling Member#principal will always work.
On the other hand calling #user will fail if
the member is a group.

machisuji committed:
use project identifier in members path

machisuji committed:
fixed members_controller spec

adapted spec to new non-js behaviour

machisuji committed:
removed signup screen; show OmniAuth opts in register view

machisuji committed:
explain why auth source wasn't deleted...

machisuji committed:
dummy LDAP auth source

I don't care if we have a company LDAP.
My thoughts go out to devs* who don't have that.

*OpenSource, right?

machisuji committed:
feature spec for user creation and activation

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://machisuji-feature-invitations.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://machisuji-feature-invitations.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

machisuji committed:
actually unsubscribe cleared subscriptions

oliverguenther commented on app/assets/javascripts/admin_users.js:
```diff
@@ -44,14 +44,21 @@
       passwordInputs.removeAttr('disabled');
+
+      newUserLogin.hide();
+      newUserLogin.find('input').prop('disabled', 'disabled');
```
Prefer to use `prop(String, boolean)` for these things, i.e., `.prop('disabled', true);`
To remove it, use `removeProp('disabled')` or `prop('disabled', false)` rather than `removeAttr` both below and above this comment.

http://api.jquery.com/prop/#prop-propertyName-value

oliverguenther commented on app/controllers/account_controller.rb:
```diff
@@ -115,11 +115,25 @@ def register
+      if session.include? :invitation_token
+        token = Token.find_by(value: session[:invitation_token])
+
+        @user = token.user
```
You have the same code for building the user from token in both blocks.
Why not move that out of the request condition?

``` ruby
if session.include? :invitation_token
  ...
elsif request.get?
  ...
else
  ... 
end
```

oliverguenther commented on app/models/dummy_auth_source.rb:
```diff
@@ -0,0 +1,39 @@
+  end
+
+  def auth_method_name
+    'DerpLAP'
```
DerpLAP, sounds promising. Where can I get me some of this cool new tech? :horse:

oliverguenther commented on app/models/user.rb:
```diff
@@ -338,6 +358,22 @@ def register!
+    update_attribute(:status, STATUSES[:invited])
+  end
+
+  def registered?
```
Mind these comments.

oliverguenther commented on app/views/account/_auth_providers.html.erb:
```diff
@@ -39,12 +39,13 @@ See doc/COPYRIGHT.rdoc for more details.
     <% unless no_pwd %>
       <h3 class="login-auth-providers-title"><span>
-          <%= I18n.t('account.login_with_auth_provider')%>
+          <%= local_assigns[:omniauth_title] || I18n.t('account.login_with_auth_provider')%>
```
Always thought `local_assigns` was considered a hack outside rails internals, but that doesn't seem to be the case anymore, since i has received documentation in the meantime. (http://nithinbekal.com/posts/rails-optional-locals/)

oliverguenther commented on app/views/invitations/index.html.erb:
```diff
@@ -0,0 +1,44 @@
+
+++#%>
+
+<% email ||= local_assigns[:email] %>
```
Wat? Either someone passed `:email` in locals, then it gets assgined to `email` anyway, or this statement will receive `nil ||= nil`. Or am I forgetting something here?

machisuji replied:
Yeah it's just to 'declare' the variable `email`, really. Cause if I don't do this and the local is not passed it will be undefined as opposed to just nil. I could also just use the `local_assigns` hash directly as you saw above.

oliverguenther replied:
I see that you'll get an error using `email` directly when it is not passed to locals, but I'd really prefer just using `local_assigns` in that case. (especially now that I know its not some internal unstable stuff).

oliverguenther commented on app/views/invitations/show.html.erb:
```diff
@@ -0,0 +1,39 @@
+</p>
+<p>
+  The created token is:
+  <a href="/invitations/claim/<%= token %>">
```
Please add an explanatory `title=""` to all links.

machisuji replied:
Mh good point. Dunno what this file is still doing in my PR, though. :D I'll delete the invitations views as that's a relict.

oliverguenther commented on app/views/members/autocomplete_for_member.json.erb:
```diff
@@ -27,16 +27,31 @@ See doc/COPYRIGHT.rdoc for more details.
   "results":
   {
+    <%# Offer to invite via email unless it's one of the found user's addresses already. #%>
+    <% email = params[:q] =~ /.+@.+/ && @principals.none? { |p| p.mail == params[:q] } %>
```
The regex probably works due to they way we restrict our logins, but I'd like to see some more explicit regex here for.

Something like `/\A\S+@\S+\.\S+\z/` is vague enough that we don't create false negatives, but  at least I cannot create an invite without TLD to `deine@muddi`

[There's lots of tradeoffs in Regex mail validation](http://www.regular-expressions.info/email.html), and thus just the basic requirements (something@something.something) would be fine for me. (Edit: See, that's how Github does it!1 :horse:)

machisuji replied:
Then perfectly valid [email addresses](https://en.wikipedia.org/wiki/Email_address#Valid_email_addresses) will be not supported, though! 

> admin@mailserver1 (local domain name with no TLD)

But yes, I suppose that's exceedingly rare.

oliverguenther replied:
full of tradeoffs.. :laughing:

oliverguenther commented on app/views/members/index.html.erb:
```diff
@@ -0,0 +1,148 @@
+                <div class="generic-table--sort-header-outer">
+                  <div class="generic-table--sort-header">
+                    <span>
+                      <%= User.model_name.human %> / <%= Group.model_name.human %>
```
`/`?

machisuji replied:
Don't ask me! I just copied this from the members settings tab view. :D

machisuji replied:
Or extracted it rather.

oliverguenther replied:
Forget my comment, I somehow overlooked the `%> <%=` parts, expecting `/` to be some operand.. d'oh.

oliverguenther commented on config/locales/en.yml:
```diff
@@ -52,6 +53,10 @@ en:
       User registration is disabled on this system. Please ask an administrator to create an
       account for you.
     login_with_auth_provider: "or sign in with your existing account"
+    signup_with_auth_provider: "or sign up using"
```
You might want to change that to include the auth provider name variable, because the translation may be more explicit then.

machisuji replied:
It's just used as the 'title' for the divider, though. The actual providers are listed below that.

oliverguenther replied:
Ah, well - that's fine for me then.

oliverguenther commented on config/locales/en.yml:
```diff
@@ -1187,6 +1196,7 @@ en:
   notice_unable_delete_version: "Unable to delete version."
   notice_automatic_set_of_standard_type: "Set standard type automatically."
   notice_logged_out: "You have been logged out."
+  notice_wont_delete_auth_source: The Authentication Mode cannot be deleted as long as there are still users using it.
```
don't capitalize Authentication Mode

oliverguenther commented on config/locales/en.yml:
```diff
@@ -1797,6 +1812,7 @@ en:
     no_login: "This user authenticates through login by password. Since it is disabled, they cannot log in."
     password_change_unsupported: Change of password is not supported.
     authorization_rejected: "You are not allowed to sign in."
+    invite: Invite user via e-mail
```
You've used `email` and `e-mail`, I think [both are fine in english](http://dictionary.cambridge.org/dictionary/english/email), but keep it consistent.

machisuji replied:
Aye.

oliverguenther commented on spec/factories/auth_source_factory.rb:
```diff
@@ -36,4 +36,8 @@
   end
+
+  factory :dummy_auth_source, class: DummyAuthSource do
+    name 'DerpLAP'
```
I suggest you let the dummy auth source reside in the specs, i.e. create it in this Factory. It shouldn't be an actual model if  we can avoid it.

machisuji replied:
I like the dummy auth source, though. I needed it for testing the use case that there is an LDAP auth source without actually having a LDAP server available. I know we have an LDAP in the company but it's just a pain because I have to lookup the data and have to connect to the VPN and stuff. Also I can only login myself as opposed to random users for testing.

Considering this and the fact that random OpenSource developers, who may want to work on OpenProject, probably don't have an LDAP either, I suggest we keep this around to make everyone's life a little bit easier.

oliverguenther replied:
I'm all for easier auth source development, but there's the requirement for testing a mocked source in the specs and then some dummy auth source for development. The latter could also easily be extracted as a plugin.

The sole problem / question for me is: Can we deactivate the auth source by default, so that actual users of OP aren't confused by that option.

machisuji replied:
You have to manually run a rake task (`rake ldap:create_dummy`) to activate this auth source.

oliverguenther replied:
Okay, that's perfectly fine for me then.

machisuji committed:
the session is optional

machisuji committed:
adapted users_controller spec to reflect new behaviour

machisuji committed:
dropped membership.feature; rewrote as adapted feature spec

The cuke was broken due to changes in the members view.
Writing and adapting cukes is a pain, hence I removed the
cuke and replaced it by a feature spec using page objects.

machisuji committed:
removed edit label (icon suffices)

machisuji committed:
fixed group memberships feature (as feature spec)

dropped cuke, dropped redundant scenarios, fixed remaining ones

machisuji committed:
replaced groups cuke by feature spec

machisuji committed:
whitespace

machisuji committed:
run without js where possible

machisuji committed:
have select support impaired mode too

machisuji committed:
fixed error messages feature (as feature spec)

- dropped cuke
- fixed a bug inviting Anonymous when no user selected
- dropped obsolete test

machisuji committed:
fixed members/members feature (as spec)

- dropped cuke
- replaced with feature spec
- DRY'd the heck out of the cuke (shared_examples ftw?)

machisuji committed:
fixed members/pagination feature (as spec)

machisuji committed:
extracted member's roles feature spec from cuke

fixed bug while at it (no error message when choosing no role)

machisuji committed:
renamed file hoping this will fix travis

machisuji committed:
fixed self registration; manual activation of invited users

machisuji committed:
polish considering remarks through review and houndci

machisuji committed:
simplified page module structure

machisuji committed:
fixed class name

oliverguenther commented on app/controllers/account_controller.rb:
```diff
@@ -153,17 +161,70 @@ def allow_lost_password_recovery?
+    end
+  end
+
+  def activate_invited(token)
```
This is some heavily branched method. Can you take these apart so that we have no more than two nesting levels? Hard for me to follow the flow here.

oliverguenther commented:
Feature LGTM :+1: , waiting for the rebase and the smaller issues we've talked about: 
- Member form: Inviting a user works without setting a role
- Consistent Spelling of E-mail
- Delete first / last name from register form when user is invited
- view-member permission
- Refactor `activate_invited`

machisuji committed:
refactoring

machisuji committed:
don't invite members without given roll

machisuji committed:
register: clear placeholder name(s); refactoring

machisuji committed:
agree on one spelling for email

machisuji committed:
only admins may invite new users

machisuji committed:
delete invited users when removed from project

machisuji committed:
auto focus mail in simple new user form

machisuji committed:
added :view_members permission

machisuji committed:
delete invited users when removed as member

machisuji committed:
groups are never disposable

machisuji committed:
moved invitation email sub to initializer

machisuji committed:
clear subscription spec

machisuji committed:
fixed specs

machisuji committed:
adapted/fixed labels

machisuji committed:
swapped icons - bin for deletion, cross for removal

machisuji committed:
show user name in deletion notices

machisuji committed:
don't offer to invite who is already a member

machisuji committed:
adapted specs to reflect latest i18n change

oliverguenther commented:
LGTM :+1: Merging this now, however keep in mind that the new `view_members` permission has not been assigned to permissions yet, you should clarify which roles apply for this permission.

machisuji commented:
@oliverguenther I did add it to the [members role](https://github.com/opf/openproject/pull/3480/files#diff-abde29a7cb3e60f19d05969edb6f2d01R88) and it's implicit for the project admin role as it were.
